GH-53: Add localization support (Resolves #53, #77) (#85)
inclusive-design/guide.inclusivedesign.ca: CONTRIBUTING.md

# Contributing

## Licensing

Code contributions ard licensed under the [3-Clause BSD](https://opensource.org/licenses/BSD-3-Clause) license.

Content contributions are licensed under the
[Creative Commons CC BY 3.0](https://creativecommons.org/licenses/by/3.0/) license.

## Process/Workflow

The general contribution workflow is for contributors to fork the project repository, work in a branch created off of

> main, and submit a pull request against the project repo's main branch to merge their contributions.

Reviewer: Should "main" be formatted differently since it's also the name of the branch. Maybe `main` or "Main".

Author: I'll put main in backticks. I think the names are case sensitive so I'd prefer not to do the capitalization.

allows for easier parsing of the type of changes to expect within a commit.

### Linting

Files are linted, statically analyzed for style and syntax conventions. Please ensure that you run the lint task before
submitting a PR.

```bash
npm run lint
```

For more information about configuration and rules see:

* [fluid-lint-all](https://www.npmjs.com/package/fluid-lint-all)
* [eslint-config-fluid](https://www.npmjs.com/package/eslint-config-fluid)
* [stylelint-config-fluid](https://www.npmjs.com/package/stylelint-config-fluid)

## Localization/Internationalization

### Adding new locales

Steps for adding additional locales.

1. Add [configuration](#locale-configuration) for the locale under the `languages` property in `src/_data/config.json`.
2. For each directory under `src/documents` duplicate the `en-CA` directory and rename based on the new locale
   1. In the new directory rename the 11tydata file after the new directory name.
   2. [Localize](#localizing-the-content) all of the Markdown files' `title` front matter and content.
3. In `src/locales` duplicate the `fr` directory and rename based on the new locale

>    1. [Update the `messages.po` and `messages.mo`](#localizing-other-text) files in the new directory for the new

Author: I added a heading before the section that talks about the .po and .mo files. I've also made the POEDIT section a subsection of this.

      locale.
      1. Update the `Language`

>       2. Update the localized text

Reviewer: @jobara can you give a little more detail here?

### Locale Configuration

The [`src/_data/config.json`](./src/_data/config.json) file contains configuraiton for each supported locale under the
`languages` property. Each locale is keyed with the language identifier using the

> [BCP-47](https://tools.ietf.org/search/bcp47) format of language-locale e.g `"en-CA"` for Canadian English. The locale

Author: Updated to use https://www.w3.org/International/articles/language-tags/

_**NOTE**: Due to [issue #23](https://github.com/sgissinger/eleventy-plugin-i18n-gettext/issues/23) when passing in the
`key`, if it isn't plain text it will not be included in messages.js. A workaround, if a variable is being used, is to
make a call with the resolved text elsewhere. For examplethe category names, used in URLs, are explicitly called in
`src/_data/site.js`. This applies to calls made in JavaScript and templates._

The specific localizations are provided in the .po and/or .mo files locataed in the [locales](./src/locales) directory.
The strings to translate are automatically picked up when a build is generated (`npm start`) and added to
[messages.js](./src/locales/messages.js). This allows for tools like [Poedit](https://poedit.net) to easily find the
strings that need to be translated. You could also update the .po and generate the .mo files manually.

#### Using Poedit

You'll need to add a JavaScript extractor when first setting up Poedit. See
[Poedit configuration for translations extraction](https://github.com/sgissinger/eleventy-plugin-i18n-gettext/blob/HEAD/docs/Manage-translations-with-Poedit.md)
for information about setup.

> To update the .po and .mo files perform the following steps:
> 
> 1. Launch Poedit
> 2. Browse for the .po file you'd like to edit
> 3. Open the .po file
> 4. `Translations > Update from Source Code`
>    * This will update the .po file with all of the current translations
>    * If this step doesn't pick up the localizations from messages.js, ensure that the source paths are set corrctly.
>      * `Translations > Properties... > Source Paths
> 5. Select the "Source Text" to localize, add the necessary "Translation"
> 6. To remove old strings go to: `Translations > Purge Deleted Translations`
> 7. Save
> 8. Make sure to commit the updates .po and .mo files to the repo.

Reviewer: ```suggestion
To update the `.po` and `.mo` files perform the following steps:

1. Launch Poedit.
2. Browse and open the `.po` file to edit.
3. Select `Translations > Update from Source Code`.
   * This will update the `.po` file with all of the current translations.
   * If this step does not find the available localizations from the `./src/locales/messages.js` file, ensure that the source paths are set correctly: `Translations > Properties... > Source Paths`.
5. Select the "Source Text" to localize, add the necessary "Translation".
6. To remove old strings go to: `Translations > Purge Deleted Translations`.
7. Save.
8. Make sure to commit the updates `.po` and `.mo` files to the working repository.
```

* [fluid-lint-all](https://www.npmjs.com/package/fluid-lint-all)
* [eslint-config-fluid](https://www.npmjs.com/package/eslint-config-fluid)
* [stylelint-config-fluid](https://www.npmjs.com/package/stylelint-config-fluid)

## Localization/Internationalization

### Adding new locales

Steps for adding additional locales.

1. Add [configuration](#locale-configuration) for the locale under the `languages` property in `src/_data/config.json`.
2. For each directory under `src/documents` duplicate the `en-CA` directory and rename based on the new locale:
   1. In the new directory rename the `.11tydata` file after the new directory name.
   2. [Localize](#localizing-the-content) all of the Markdown files' `title` front matter and content.
3. In `src/locales` duplicate the `fr` directory and rename based on the new locale
   1. [Update `messages.po` and generate a new `messages.mo`](#localized-strings) file in the new directory for the new
      locale.
      1. Update the `Language`
      2. Update the localized text
      3. Using a tool souch as [Poedit](https://poedit.net), generate a new `messages.mo` file from `messages.po`.

### Locale Configuration

> The [`src/_data/config.json`](./src/_data/config.json) file contains configuraiton for each supported locale under the

Reviewer: ```suggestion
The [`src/_data/config.json`](./src/_data/config.json) file contains configuration for each supported locale under the
```

[BCP-47 syntax](https://www.w3.org/International/articles/language-tags/) of language-locale e.g `"en-CA"` for Canadian
English. The locale must specify the following:

* **slug**: used in the URL to display the localized resource. This must be unique and match the locale being
  configured. It could be in the form of just language (e.g. `"en"`) or language-locale (e.g. `"en-CA"` or `"en-ca"`).
  For the default language, the locale slug will not be added to the URL.
* **uioSlug**: passed to UI Options widget to specify its locale to use. UI Options uses the form language (e.g `"en"`)
  or language_locale (e.g. `"en_CA"`). UI Options provides its own
  [localization](https://github.com/fluid-project/infusion/tree/main/src/framework/preferences/messages) outside of the
  context of this site and will fall back to English if it does not support a requested localization.
* **dir**: The direction of the language, either `"ltr"` (left-to-right) or `"rtl"` (right-to-left). It will default to
  `"ltr"` if not specified.
* **name**: The human readable name of the language, in the locale's language. It is used to identify which languages

>   are availalbe to switch between.

Reviewer: ```suggestion
  are available to switch between.
```

For example, to add a France-French locale, the following is added to the `languages` block in
`src/_data/config.json`:

```JSON
{
    "fr-FR": {
        "slug": "fr-FR",
        "uioSlug": "fr-FR",
        "dir": "ltr",
        "name": "Français (France)"
    }
}
```

### Localizing the content

The main content of the site is all contained within the [`src/documents`](./src/documents) directory. These are
separated into collections/directories for the various categories (i.e. `activities/`, `insights/`, `tools/`,
`practices/`) and pages (i.e. `index.md`, `404.md` in for each language in the `pages` directory).

> Each collection is split into sub-directories cooresponding to the locales. Each locale dirctory will contain the

Reviewer: ```suggestion
Each collection is split into sub-directories coresponding to the locales. Each locale directory will contain the
```

collection; with the content of the files being localized as needed.

For example if adding a new page to the Activities called "My Activity" we'd add a `MyActivity.md` file to each
locale. e.g. `src/documents/activities/en-CA/MyActivity.md` and `src/documents/activities/fr-FR/MyActivity.md`. They
both retain the same file name, which allows for switching between the localized version with the language switcher on
the site.

The English (`en-CA`) version might look like.

```Markdown
---
title: My Activity
eleventyNavigation:
    parent: Activities
    key: My Activity
---

Content for My Activity.

```

The French (`fr-FR`) version might look like.

```Markdown
---
title: Mon activité
eleventyNavigation:
    parent: Activities
    key: My Activity
---

Contenu pour Mon Activité.
```

It's important to note that only two parts are localized; the main content portion of the Markdown file, and the `title`
in the YAML front matter. Other portions of the front matter is likely not localized as it is often used for
configuration as opposed to content. In the above example, the `eleventyNavigation` configures the category page
navigation.

> Additionly each, locale directory will contain an

Reviewer: ```suggestion
Additionally each, locale directory will contain an
```

<h1>Example</h1>

<p>Using Nunjucks template engine</p>
```

A French localization might render like:

```HTML
<h1>Exemple</h1>

<p>Utiliser le moteur de template Nunjucks</p>
```

In JavaScript, import the [eleventy-plugin-i18n-gettext](https://www.npmjs.com/package/eleventy-plugin-i18n-gettext) and
make calls to the functions.

```JavaScript
"use strict";

const i18n = require("eleventy-plugin-i18n-gettext");

module.exports = (data) => {
    const locale = "en-CA";
    return i18n._(locale, "To translate");
};
```

_**NOTE**: locale variables need to be prefixed with `locale_` due to [eleventy-plugin-i18n-gettext issue #22](https://github.com/sgissinger/eleventy-plugin-i18n-gettext/issues/22)._

_**NOTE**: Ensure that `key`  is plain text when using gettext function calls, otherwise it will not be included in the
`messages.js` file which is used by .PO editors.  See
[eleventy-plugin-i18n-gettext issue #23](https://github.com/sgissinger/eleventy-plugin-i18n-gettext/issues/23). As a
workaround for JavaScript and templates, resolve the `key` variable elsewhere. For example the category names used in
URLs are explicitly called in `src/_data/site.js`._

##### Localized strings

> The specific localization strings are provided in the `.po` and/or `.mo` files locataed in the

Reviewer: ```suggestion
The specific localization strings are provided in the `.po` and/or `.mo` files located in the
```

submitting a PR.

```bash
npm run lint
```

For more information about configuration and rules see:

* [fluid-lint-all](https://www.npmjs.com/package/fluid-lint-all)
* [eslint-config-fluid](https://www.npmjs.com/package/eslint-config-fluid)
* [stylelint-config-fluid](https://www.npmjs.com/package/stylelint-config-fluid)

## Localization/Internationalization

### Adding new locales

Steps for adding additional locales.

1. Add [configuration](#locale-configuration) for the locale under the `languages` property in `src/_data/config.json`.
2. For each directory under `src/documents` duplicate the `en-CA` directory and rename based on the new locale:
   1. In the new directory rename the `.11tydata` file after the new directory name.
   2. [Localize](#localizing-the-content) all of the Markdown files' `title` front matter and content.
3. In `src/locales` duplicate the `fr` directory and rename based on the new locale
   1. [Update `messages.po` and generate a new `messages.mo`](#localized-strings) file in the new directory for the new
      locale.
      1. Update the `Language`

>       2. Update the localized text by replacing the `msgstr` values in the .po file with the corresponding text from the

Reviewer: ```suggestion
      2. Update the localized text by replacing the `msgstr` values in the `.po` file with the corresponding text from the
```

>          new localization. Can use a tool such as [Poedit](https://poedit.net) to edit the .po file.

Reviewer: ```suggestion
         new localization. Can use a tool such as [Poedit](https://poedit.net) to edit the `.po` file.
```

* `{{ _i() }}` - allows for interpolation of values into the translated string. Example:
  `{{ _i("licensed under a${license}.", {license: "license"}) | safe }}`.

For example:

```Nunjucks
<h1>{{ _("Example") }}</h1>

<p>{{ _i("Using ${templateName} template engine.", {templateName: "Nunjucks"}) }}</p>
```

Would render the follwing in English:

```HTML
<h1>Example</h1>

<p>Using Nunjucks template engine</p>
```

A French localization might render like:

```HTML
<h1>Exemple</h1>

<p>Utiliser le moteur de template Nunjucks</p>
```

In JavaScript, import the [eleventy-plugin-i18n-gettext](https://www.npmjs.com/package/eleventy-plugin-i18n-gettext) and
make calls to the functions.

```JavaScript
"use strict";

const i18n = require("eleventy-plugin-i18n-gettext");

module.exports = (data) => {
    const locale = "en-CA";
    return i18n._(locale, "To translate");
};
```

_**NOTE**: locale variables need to be prefixed with `locale_` due to [eleventy-plugin-i18n-gettext issue #22](https://github.com/sgissinger/eleventy-plugin-i18n-gettext/issues/22)._

_**NOTE**: Ensure that `key`  is plain text when using gettext function calls, otherwise it will not be included in the

> `messages.js` file which is used by .PO editors.  See

Reviewer: ```suggestion
`messages.js` file which is used by `.po` file editors.  See
```
